File: backend/app/models/schemas/ticket.py
"""
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator

# Reuse the ORM enums so the API contract can never drift from the DB.
from app.models.investigation import InvestigationRunStatus, InvestigationRunType
from app.models.ticket import ResolutionOutcome, TicketPriority, TicketSource, TicketStatus
from app.models.ticket_activity import TicketActivityType, TicketActorType
from app.models.schemas.pagination import Pagination

MAX_TITLE_LENGTH = 500
MAX_DESCRIPTION_LENGTH = 20000
MAX_COMMENT_LENGTH = 10000
MAX_TAGS = 5


class TicketCreate(BaseModel):
    title: str = Field(min_length=1, max_length=MAX_TITLE_LENGTH)
    description: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    priority: TicketPriority = TicketPriority.MEDIUM
    severity: Optional[int] = Field(default=None, ge=1, le=3)
    tags: Optional[List[str]] = Field(default=None, max_length=MAX_TAGS)
    customer_id: Optional[UUID] = None
    # Manual tickets: resolve/create the customer by email so direct-email
    # notifications have somewhere to go. Ignored when customer_id is set.
    customer_email: Optional[EmailStr] = None
    customer_name: Optional[str] = Field(default=None, max_length=200)
    session_id: Optional[UUID] = None
    assignee_user_id: Optional[UUID] = None
    group_id: Optional[UUID] = None

    @field_validator("title")
    @classmethod
    def _strip_title(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("must not be blank")
        return v


class TicketUpdate(BaseModel):
    """Partial update — omitted fields keep their current values (apply with
    model_dump(exclude_unset=True))."""
    title: Optional[str] = Field(default=None, min_length=1, max_length=MAX_TITLE_LENGTH)
    description: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    # Set/replace the ticket's customer by email (find-or-create in the org).
    customer_email: Optional[EmailStr] = None
    customer_name: Optional[str] = Field(default=None, max_length=200)
    status: Optional[TicketStatus] = None
    priority: Optional[TicketPriority] = None
    severity: Optional[int] = Field(default=None, ge=1, le=3)
    tags: Optional[List[str]] = Field(default=None, max_length=MAX_TAGS)
    assignee_user_id: Optional[UUID] = None
    group_id: Optional[UUID] = None
    resolution_outcome: Optional[ResolutionOutcome] = None
    resolution_summary: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    customer_resolution_message: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)


class TicketCommentCreate(BaseModel):
    body: str = Field(min_length=1, max_length=MAX_COMMENT_LENGTH)
    is_internal: bool = True

    @field_validator("body")
    @classmethod
    def _strip_body(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("must not be blank")
        return v


class TicketReopen(BaseModel):
    reason: Optional[str] = Field(default=None, max_length=MAX_COMMENT_LENGTH)


class TicketResolve(BaseModel):
    outcome: ResolutionOutcome = ResolutionOutcome.FIXED
    resolution_summary: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    customer_message: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)


class TicketUserOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    full_name: Optional[str] = None
    email: Optional[str] = None


class TicketCustomerOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    email: Optional[str] = None
    full_name: Optional[str] = None


class TicketActivityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    activity_type: TicketActivityType
    actor_type: TicketActorType
    actor_user_id: Optional[UUID] = None
    actor_name: Optional[str] = None
    body: Optional[str] = None
    is_internal: bool = True
    activity_metadata: Optional[dict] = None
    created_at: Optional[datetime] = None


class InvestigationRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    run_type: InvestigationRunType
    status: InvestigationRunStatus
    trigger: str
    error: Optional[str] = None
    tool_calls_used: int = 0
    max_tool_calls: int = 25
    model_name: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: Optional[datetime] = None


class HypothesisOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    idx: int
    title: str
    rationale: Optional[str] = None
    status: str
    confidence: Optional[float] = None
    conclusion: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class InvestigationEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    hypothesis_id: Optional[UUID] = None
    seq: int
    event_type: str
    label: Optional[str] = None
    tool_name: Optional[str] = None
    connector_name: Optional[str] = None
    tool_input: Optional[str] = None
    tool_result: Optional[str] = None
    duration_ms: Optional[int] = None
    error: Optional[str] = None
    created_at: Optional[datetime] = None


class RCADocumentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    run_id: Optional[UUID] = None
    version: int
    summary: Optional[str] = None
    impact: Optional[str] = None
    timeline: Optional[list] = None
    investigation_log: Optional[str] = None
    contributing_factors: Optional[list] = None
    conclusion: Optional[str] = None
    remediation: Optional[str] = None
    prevention: Optional[str] = None
    customer_summary: Optional[str] = None
    confidence: Optional[float] = None
    is_partial: bool = False
    generated_by: str = "ai"
    reviewed_by_user_id: Optional[UUID] = None
    reviewed_by_name: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ProposalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    run_id: Optional[UUID] = None
    summary: str
    customer_message: Optional[str] = None
    confidence: Optional[float] = None
    status: str
    decided_by_user_id: Optional[UUID] = None
    decided_by_name: Optional[str] = None
    decided_at: Optional[datetime] = None
    reject_reason: Optional[str] = None
    created_at: Optional[datetime] = None


class ProposalRejectRequest(BaseModel):
    reason: Optional[str] = Field(default=None, max_length=MAX_COMMENT_LENGTH)
    # Re-enqueue a refined investigation with the reason as context.
    reinvestigate: bool = False


class InvestigationDetailResponse(BaseModel):
    """Glass-box payload for the ticket detail page: the (latest or requested)
    investigation run with its hypotheses, evidence events and RCA document."""
    run: Optional[InvestigationRunOut] = None
    hypotheses: List[HypothesisOut] = []
    events: List[InvestigationEventOut] = []
    rca: Optional[RCADocumentOut] = None
    # Latest proposal (any status) — drives the approval banner.
    proposal: Optional[ProposalOut] = None


class RcaUpdate(BaseModel):
    customer_summary: Optional[str] = Field(default=None, max_length=MAX_DESCRIPTION_LENGTH)
    # Stamp the caller as reviewer ("generated by AI · reviewed by …").
    mark_reviewed: bool = False


class InvestigateRequest(BaseModel):
    run_type: InvestigationRunType = InvestigationRunType.INVESTIGATION
    # Free-text context for the run, e.g. what to focus on this time.
    context_note: Optional[str] = Field(default=None, max_length=4000)


class TicketOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    ticket_number: int
    display_number: str
    organization_id: UUID
    customer_id: Optional[UUID] = None
    title: str
    original_title: Optional[str] = None
    description: Optional[str] = None
    status: TicketStatus
    priority: TicketPriority
    severity: Optional[int] = None
    source: TicketSource
    intent: Optional[str] = None
    triage_confidence: Optional[float] = None
    ai_summary: Optional[str] = None
    tags: Optional[List[str]] = None
    assignee_user_id: Optional[UUID] = None
    group_id: Optional[UUID] = None
    agent_id: Optional[UUID] = None
    duplicate_of_ticket_id: Optional[UUID] = None
    resolution_outcome: Optional[ResolutionOutcome] = None
    resolution_summary: Optional[str] = None
    customer_resolution_message: Optional[str] = None
    first_response_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    confirmation_requested_at: Optional[datetime] = None
    reopened_count: int = 0
    csat_requested_at: Optional[datetime] = None
    csat_score: Optional[int] = None
    csat_responded_at: Optional[datetime] = None
    external_ref_type: Optional[str] = None
    external_ref_id: Optional[str] = None
    external_ref_url: Optional[str] = None
    created_by_user_id: Optional[UUID] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Derived UI fields.
    ai_state: Optional[str] = None
    sla_due_at: Optional[datetime] = None
    assignee: Optional[TicketUserOut] = None
    customer: Optional[TicketCustomerOut] = None


class TicketListItem(BaseModel):
    """Slim row for the list view."""
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    ticket_number: int
    display_number: str
    title: str
    status: TicketStatus
    priority: TicketPriority
    tags: Optional[List[str]] = None
    assignee_user_id: Optional[UUID] = None
    assignee_name: Optional[str] = None
    ai_state: Optional[str] = None
    sla_due_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class TicketListResponse(BaseModel):
    tickets: List[TicketListItem]
    pagination: Pagination


class TicketDetailResponse(BaseModel):
    ticket: TicketOut
    activities: List[TicketActivityOut]
    runs: List[InvestigationRunOut]
    linked_session_ids: List[UUID]
    possible_duplicates: List[TicketListItem] = []
    # Whether any outbound path to the customer exists (linked conversation or
    # customer email) — gates the "send to customer" affordances in the UI.
    can_notify_customer: bool = False


class TicketStats(BaseModel):
    open: int
    awaiting_approval: int
    sla_breaching: int
    ai_resolved_pct_7d: Optional[float] = None
    # CSAT over the trailing csat_window_days, overall and split by resolver.
    csat_avg: Optional[float] = None
    csat_ai_avg: Optional[float] = None
    csat_human_avg: Optional[float] = None
    csat_responses: int = 0
    csat_window_days: int = 30


class SlaTarget(BaseModel):
    first_response_minutes: int = Field(gt=0, le=60 * 24 * 30)
    resolution_minutes: int = Field(gt=0, le=60 * 24 * 90)


class TicketSettingsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    autonomy_level: int
    auto_investigate_on_create: bool
    min_confidence_to_auto_resolve: float
    confirmation_timeout_hours: int
    csat_enabled: bool
    sla_targets: Optional[dict] = None
    created_template: Optional[str] = None
    resolved_template: Optional[str] = None
    jira_escalation_enabled: bool
    jira_escalation_priority: Optional[str] = None
    investigation_mcp_tool_ids: Optional[List[int]] = None
    alert_webhook_enabled: bool
    # Only exposed to settings managers; forms the intake URL path.
    alert_webhook_secret: Optional[str] = None
    max_tool_calls_per_run: int
    max_runs_per_ticket: int


class TicketSettingsUpdate(BaseModel):
    autonomy_level: Optional[int] = Field(default=None, ge=1, le=3)
    auto_investigate_on_create: Optional[bool] = None
    min_confidence_to_auto_resolve: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    confirmation_timeout_hours: Optional[int] = Field(default=None, ge=1, le=24 * 30)
    csat_enabled: Optional[bool] = None
    sla_targets: Optional[dict] = None
    created_template: Optional[str] = Field(default=None, max_length=2000)
    resolved_template: Optional[str] = Field(default=None, max_length=2000)
    jira_escalation_enabled: Optional[bool] = None
    jira_escalation_priority: Optional[TicketPriority] = None
    investigation_mcp_tool_ids: Optional[List[int]] = None
    alert_webhook_enabled: Optional[bool] = None
    max_tool_calls_per_run: Optional[int] = Field(default=None, ge=1, le=100)
    max_runs_per_ticket: Optional[int] = Field(default=None, ge=1, le=10)

    @field_validator("sla_targets")
    @classmethod
    def _validate_sla(cls, v: Optional[dict]) -> Optional[dict]:
        if v is None:
            return v
        for priority, target in v.items():
            if priority not in [p.value for p in TicketPriority]:
                raise ValueError(f"unknown priority '{priority}'")
            SlaTarget.model_validate(target)
        return v

File: frontend/src/types/ticket.ts
/*
Copyright 2024-2026 ChatterMate

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

export type TicketStatus =
  | 'open'
  | 'triaging'
  | 'investigating'
  | 'awaiting_approval'
  | 'in_progress'
  | 'resolved_pending_confirmation'
  | 'resolved'
  | 'closed'
  | 'reopened'

export type TicketPriority = 'urgent' | 'high' | 'medium' | 'low'

export type TicketAiState = 'investigating' | 'awaiting' | 'human' | 'resolved'

export type ResolutionOutcome =
  | 'fixed'
  | 'workaround'
  | 'not_a_bug'
  | 'duplicate'
  | 'cannot_reproduce'
  | 'customer_unresponsive'
  | 'escalated_external'

export interface TicketListItem {
  id: string
  ticket_number: number
  display_number: string
  title: string
  status: TicketStatus
  priority: TicketPriority
  tags?: string[] | null
  assignee_user_id?: string | null
  assignee_name?: string | null
  ai_state?: TicketAiState | null
  sla_due_at?: string | null
  resolved_at?: string | null
  created_at?: string | null
  updated_at?: string | null
}

export interface Ticket extends TicketListItem {
  organization_id: string
  customer_id?: string | null
  original_title?: string | null
  description?: string | null
  severity?: number | null
  source: 'chat_ai' | 'human_agent' | 'manual' | 'api'
  intent?: string | null
  triage_confidence?: number | null
  ai_summary?: string | null
  group_id?: string | null
  agent_id?: string | null
  duplicate_of_ticket_id?: string | null
  resolution_outcome?: ResolutionOutcome | null
  resolution_summary?: string | null
  customer_resolution_message?: string | null
  first_response_at?: string | null
  closed_at?: string | null
  confirmation_requested_at?: string | null
  reopened_count: number
  external_ref_type?: string | null
  external_ref_id?: string | null
  external_ref_url?: string | null
  created_by_user_id?: string | null
  assignee?: { id: string; full_name?: string | null; email?: string | null } | null
  customer?: { id: string; full_name?: string | null; email?: string | null } | null
}

export type TicketActivityType =
  | 'comment'
  | 'status_change'
  | 'assignment'
  | 'priority_change'
  | 'ai_triage'
  | 'ai_investigation_started'
  | 'ai_investigation_completed'
  | 'ai_resolution_proposed'
  | 'ai_resolution_approved'
  | 'ai_resolution_rejected'
  | 'customer_notified'
  | 'customer_replied'
  | 'customer_linked'
  | 'csat_requested'
  | 'reopened'
  | 'jira_escalated'

export interface TicketActivity {
  id: string
  activity_type: TicketActivityType
  actor_type: 'user' | 'ai' | 'customer' | 'system'
  actor_user_id?: string | null
  actor_name?: string | null
  body?: string | null
  is_internal: boolean
  activity_metadata?: Record<string, any> | null
  created_at?: string | null
}

export interface InvestigationRun {
  id: string
  run_type: 'triage' | 'investigation'
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'budget_exceeded'
  trigger: string
  error?: string | null
  tool_calls_used: number
  max_tool_calls?: number
  model_name?: string | null
  started_at?: string | null
  finished_at?: string | null
  created_at?: string | null
}

export type HypothesisStatus = 'pending' | 'testing' | 'validated' | 'invalidated' | 'inconclusive'

export interface InvestigationHypothesis {
  id: string
  idx: number
  title: string
  rationale?: string | null
  status: HypothesisStatus
  confidence?: number | null
  conclusion?: string | null
  created_at?: string | null
  updated_at?: string | null
}

export interface InvestigationEvent {
  id: string
  hypothesis_id?: string | null
  seq: number
  event_type: 'phase' | 'tool_call'
  label?: string | null
  tool_name?: string | null
  connector_name?: string | null
  tool_input?: string | null
  tool_result?: string | null
  duration_ms?: number | null
  error?: string | null
  created_at?: string | null
}

export interface RcaTimelineEntry {
  time: string
  event: string
}

export interface RcaDocument {
  id: string
  run_id?: string | null
  version: number
  summary?: string | null
  impact?: string | null
  timeline?: RcaTimelineEntry[] | null
  investigation_log?: string | null
  contributing_factors?: string[] | null
  conclusion?: string | null
  remediation?: string | null
  prevention?: string | null
  customer_summary?: string | null
  confidence?: number | null
  is_partial: boolean
  generated_by: string
  reviewed_by_user_id?: string | null
  reviewed_by_name?: string | null
  reviewed_at?: string | null
  created_at?: string | null
  updated_at?: string | null
}

export type ProposalStatus = 'pending' | 'approved' | 'rejected' | 'superseded'

export interface TicketProposal {
  id: string
  run_id?: string | null
  summary: string
  customer_message?: string | null
  confidence?: number | null
  status: ProposalStatus
  decided_by_user_id?: string | null
  decided_by_name?: string | null
  decided_at?: string | null
  reject_reason?: string | null
  created_at?: string | null
}

/** Glass-box payload for the ticket detail page. */
export interface InvestigationDetail {
  run: InvestigationRun | null
  hypotheses: InvestigationHypothesis[]
  events: InvestigationEvent[]
  rca: RcaDocument | null
  proposal: TicketProposal | null
}

export interface DbConnectorColumn {
  name: string
  type: string
}

export interface DbConnectorTable {
  schema: string
  table: string
  columns: DbConnectorColumn[]
}

export interface DbConnector {
  id: string
  name: string
  engine: 'postgresql' | 'mysql'
  host: string
  port: number
  database: string
  username: string
  enabled: boolean
  allowed_tables?: string[] | null
  masked_columns?: string[] | null
  // { "schema.table": "customer_column" } — listed tables return only the
  // ticket customer's own rows.
  row_scope?: Record<string, string> | null
  row_scope_key?: 'email' | 'phone'
  max_rows: number
  statement_timeout_ms: number
  ssh_enabled: boolean
  ssh_host?: string | null
  ssh_port: number
  ssh_username?: string | null
  last_test_at?: string | null
  last_test_ok?: boolean | null
  created_at?: string | null
}

export interface DbConnectorPayload {
  name: string
  engine: 'postgresql' | 'mysql'
  host: string
  port: number
  database: string
  username: string
  password?: string
  enabled?: boolean
  allowed_tables?: string[]
  masked_columns?: string[]
  row_scope?: Record<string, string>
  row_scope_key?: 'email' | 'phone'
  max_rows?: number
  statement_timeout_ms?: number
  // SSH tunnel (bastion/jump host) — secrets are write-only.
  ssh_enabled?: boolean
  ssh_host?: string
  ssh_port?: number
  ssh_username?: string
  ssh_password?: string
  ssh_private_key?: string
  ssh_private_key_passphrase?: string
}

export interface TicketDetail {
  ticket: Ticket
  activities: TicketActivity[]
  runs: InvestigationRun[]
  linked_session_ids: string[]
  possible_duplicates?: TicketListItem[]
  /** Linked conversation or customer email exists — gates "send to customer" UI. */
  can_notify_customer?: boolean
}

export interface Pagination {
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface TicketListResponse {
  tickets: TicketListItem[]
  pagination: Pagination
}

export interface TicketStats {
  open: number
  awaiting_approval: number
  sla_breaching: number
  ai_resolved_pct_7d?: number | null
}

export interface TicketListFilters {
  status: string
  priority: string
  assignee: string
  ai: string
  search: string
  sort: string
}

export interface TicketCreatePayload {
  title: string
  description?: string
  priority?: TicketPriority
  severity?: number
  tags?: string[]
  customer_id?: string
  /** Manual tickets: find-or-create the customer by email for direct-email notifications. */
  customer_email?: string
  customer_name?: string
  session_id?: string
  assignee_user_id?: string
  group_id?: string
}

export interface TicketUpdatePayload {
  title?: string
  description?: string
  /** Set/replace the ticket's customer by email (find-or-create in the org). */
  customer_email?: string
  customer_name?: string
  status?: TicketStatus
  priority?: TicketPriority
  severity?: number
  tags?: string[]
  assignee_user_id?: string | null
  group_id?: string | null
  resolution_outcome?: ResolutionOutcome
  resolution_summary?: string
  customer_resolution_message?: string
}

export interface SlaTarget {
  first_response_minutes: number
  resolution_minutes: number
}

export interface TicketSettings {
  autonomy_level: number
  auto_investigate_on_create: boolean
  min_confidence_to_auto_resolve: number
  confirmation_timeout_hours: number
  csat_enabled: boolean
  sla_targets?: Record<TicketPriority, SlaTarget> | null
  created_template?: string | null
  resolved_template?: string | null
  jira_escalation_enabled: boolean
  jira_escalation_priority?: TicketPriority | null
  investigation_mcp_tool_ids?: number[] | null
  alert_webhook_enabled: boolean
  alert_webhook_secret?: string | null
  max_tool_calls_per_run: number
  max_runs_per_ticket: number
}

export interface TicketUpdateEvent {
  ticket_id: string
  kind: 'created' | 'status' | 'triage' | 'comment' | 'run' | 'updated'
  payload: Record<string, any>
}
